broadcaster/_backends/redis.py

```diff
@@ -1,34 +1,99 @@
+        self._mpsc: typing.Optional[Receiver] = None
 
     async def connect(self) -> None:
-        self._pub_conn = await asyncio_redis.Connection.create(self._host, self._port)
-        self._sub_conn = await asyncio_redis.Connection.create(self._host, self._port)
-        self._subscriber = await self._sub_conn.start_subscribe()
+        if self._pub_conn or self._sub_conn or self._msg_queue:
+            logger.warning("connections are already setup but connect called again; not doing anything")
+            return
+
+        self._pub_conn = await aioredis.create_redis(self.conn_url)
+        self._sub_conn = await aioredis.create_redis(self.conn_url)
+        self._msg_queue = asyncio.Queue()  # must be created here, to get proper event loop
+        self._mpsc = Receiver()
+        self._reader_task = asyncio.create_task(self._reader())
 
     async def disconnect(self) -> None:
-        self._pub_conn.close()
-        self._sub_conn.close()
+        if self._pub_conn and self._sub_conn:
+            self._pub_conn.close()
+            self._sub_conn.close()
+        else:
+            logger.warning("connections are not setup, invalid call to disconnect")
+
+        self._pub_conn = None
+        self._sub_conn = None
+        self._msg_queue = None
+
+        if self._mpsc:
+            self._mpsc.stop()
+        else:
+            logger.warning("redis mpsc receiver is not set, cannot stop it")
+
+        if self._reader_task:
+            if self._reader_task.done():
+                self._reader_task.result()
+            else:
+                logger.debug("cancelling reader task")
+                self._reader_task.cancel()
+                self._reader_task = None
 
     async def subscribe(self, channel: str) -> None:
-        await self._subscriber.subscribe([channel])
+        if not self._sub_conn:
+            logger.error(f"not connected, cannot subscribe to channel {channel!r}")
+            return
+
+        await self._sub_conn.subscribe(self._mpsc.channel(channel))
 
     async def unsubscribe(self, channel: str) -> None:
-        await self._subscriber.unsubscribe([channel])
+        if not self._sub_conn:
+            logger.error(f"not connected, cannot unsubscribe from channel {channel!r}")
+            return
+
+        await self._sub_conn.unsubscribe(channel)
 
     async def publish(self, channel: str, message: typing.Any) -> None:
-        await self._pub_conn.publish(channel, message)
+        if not self._pub_conn:
+            logger.error(f"not connected, cannot publish to channel {channel!r}")
+            return
+
+        await self._pub_conn.publish_json(channel, message)
 
     async def next_published(self) -> Event:
-        message = await self._subscriber.next_published()
-        return Event(channel=message.channel, message=message.value)
+        if not self._msg_queue:
+            raise RuntimeError("unable to get next_published event, RedisBackend is not connected")
```

Comment:
`RuntimeError` doesn't seem correct to me... Depending on how the Broadcaster base uses this, I imagine it might be `StopIteration` or something.

Reply:
`StopIteration` is used specifically when implementing an iterator so I don't think that would be correct. This error is very unlikely to be hit when using the Broadcaster base, this is mostly just a protection if someone used this class directly for some reason.